### PR TITLE
Implement trace trimming.

### PR DESCRIPTION
**Pull request by vext01**
See the commit message for a detailed description.

A companion PR follows. This PR ought to go in first, but we have a CI cycle. We need to think about a better way to deal with these kinds of breaking changes, as I expect many more in the future.

Companion PR: https://github.com/softdevteam/ykrustc/pull/55

### CODE REVIEW

File: yktrace/src/lib.rs [hunk 1]
```diff
@@ -13,10 +13,7 @@
 extern crate test;
 
 use core::yk_swt::SirLoc;
-use std::{
```
Comment (vext01) [inline yktrace/src/lib.rs]: This hunk is due to a rustfmt change. I wasn't sure if it should be in a separate commit or not.

File: ykpack/src/types.rs [hunk 2]
```diff
@@ -108,21 +108,29 @@ impl Display for DefId {
     }
 }
 
+/// Bits in the `flags` bitfield in `Body`.
+pub mod bodyflags {
+    pub const TRACE_HEAD: u8 = 1;
+    pub const TRACE_TAIL: u8 = 1 << 1;
```
Comment (ptersilie) [inline ykpack/src/types.rs]: `pub const TRACE_TAIL: u8 = 2;`
or
`pub const TRACE_TAIL: u8 = 0x10;`
?

Comment (ptersilie) [reply ykpack/src/types.rs]: I would probably write it like this:
```
pub const TRACE_HEAD: u8 = 0x01;
pub const TRACE_TAIL: u8 = 0x10;
```
So it's immediately clear these are flags, and doesn't require any additional decoding by someone reading the code.

File: yktrace/src/lib.rs [hunk 3]
```diff
@@ -26,42 +23,70 @@ mod swt;
...
-        if self.next_idx < self.trace.len() {
-            let ret = self.trace.loc(self.next_idx);
-            self.next_idx += 1;
-            Some(ret)
+        if self.next_idx < self.trace.raw_len() {
+            let def_id = DefId::from_sir_loc(self.trace.raw_loc(self.next_idx));
+            if SIR.markers.trace_tails.contains(&def_id) {
+                // Stop when we find the start of the code that stops the tracer, thus trimming the
+                // end of the trace. By setting the next index to one above the last one in the
+                // trace, we we ensure the iterator will return `None` forever more.
```
Comment (ptersilie) [inline yktrace/src/lib.rs]: `we we` typo

File: ykpack/src/types.rs (hunk 2 above)
Comment (vext01) [reply ykpack/src/types.rs]: I hate to break it to you, but `0x10` is not 2 :P

But more seriously, the `<<` notation is more conventional, because it means you don't have to calculate the power of 2 manually. Admittedly that's not a big issue for a `u8` bitfield, but still.

File: yktrace/src/lib.rs (hunk 3 above)
Comment (vext01) [reply yktrace/src/lib.rs]: Good catch. Thanks.

File: yktrace/src/tir.rs [hunk 4]
```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
...
 pub struct TirTrace {
     ops: Vec<TirOp>
 }
 
 impl TirTrace {
-    /// Create a TirTrace from a SirTrace.
-    /// Returns Err if a DefId is encountered for which no SIR is available.
-    pub(crate) fn new<'s>(trace: &'s dyn SirTrace) -> Result<Self, InvalidTraceError> {
+    /// Create a TirTrace from a SirTrace, trimming remnants of the code which starts/stops the
+    /// tracer. Returns a TIR trace and the bounds the SIR trace was trimmed to, or Err if a DefId
```
Comment (ptersilie) [inline yktrace/src/tir.rs]: `and the(n) bounds the SIR trace` typo

File: ykpack/src/types.rs (hunk 2 above)
Comment (ptersilie) [reply ykpack/src/types.rs]: > I hate to break it to you, but 0x10 is not 2 :P
Haha, sorry, I meant `0b02`. :D

> But more seriously, the << notation is more conventional
If this is the convention then that's fair enough. I personally just find it easier to read `0b10` instead of `1 << 1`.

Comment (vext01) [reply ykpack/src/types.rs]: `0b` is fine for a `u8`, but now consider a u32:
`0b00000000100000000000000000000000` vs. `1<<23`.

Or a `u64`...

Comment (ltratt) [reply ykpack/src/types.rs]: Can a trace be both `TRACE_HEAD` and `TRACE_TAIL`? If not, then there's no need to worry about bits in quite this way: just a numeric differentiation would be enough, right?

File: yktrace/src/debug.rs [hunk 5]
```diff
@@ -48,20 +48,38 @@ pub fn def_path(def_id: &DefId) -> Option<&str> {
...
-    for loc in trace {
+    for loc in locs {
         let def_id = DefId::from_sir_loc(&loc);
         let def_path_s = match def_path(&def_id) {
             Some(s) => s,
             None => "<unknown>"
         };
+
+        let body = SIR.bodies.get(&def_id);
+        let mut flags = Vec::new();
```
Comment (ltratt) [inline yktrace/src/debug.rs]: This intermediate `Vec` feels like a real song and dance for nothing. Why not move the `if` after the `println` and just `println("HEAD")` (or `TAIL`) as appropriate?

File: yktrace/src/lib.rs [hunk 6]
```diff
@@ -207,12 +233,43 @@ mod test_helpers {
...
+    fn trim_trace() {
+        let tracer = start_tracing(Some(TracingKind::SoftwareTracing));
+        work(black_box(100));
+        let sir_trace = tracer.t_impl.stop_tracing().unwrap();
+
+        let contains_tracer_start_stop = |locs: Vec<&SirLoc>| {
+            let mut found_start_code = false;
+            let mut found_stop_code = false;
+            for loc in locs {
+                let body = match SIR.bodies.get(&DefId::from_sir_loc(&loc)) {
```
Comment (ltratt) [inline yktrace/src/lib.rs]: This can just be an `unwrap` I think?

File: yktrace/src/tir.rs [hunk 7]
```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
...
+pub struct Sir {
+    /// Lets us map a DefId from a trace to a SIR body.
+    pub bodies: HashMap<DefId, Body>,
+    // Interesting locations that we need quick access to.
+    pub markers: SirMarkers
+}
+
+/// Contains the DefIds of interesting locations required for trace manipulation.
+pub struct SirMarkers {
+    /// The function which starts tracing and whose suffix gets trimmed off the top of traces.
```
Comment (ltratt) [inline yktrace/src/tir.rs]: It's not the function's suffix that gets trimmed off, I assume, but the whole function that gets trimmed off the trace?

File: yktrace/src/tir.rs [hunk 8]
```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
...
+    /// There is only one such function: `yktrace::start_tracing()`.
+    ///
+    /// The compiler is made aware of this location by the `#[trace_head]` annotation.
+    pub trace_head: DefId,
+    /// Functions which stop tracing and whose prefix gets trimmed off the bottom of traces.
```
Comment (ltratt) [inline yktrace/src/tir.rs]: Similarly, I assume the entirety of this function is removed from the end of the trace?

File: yktrace/src/tir.rs [hunk 9]
```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
...
-    pub(crate) fn new<'s>(trace: &'s dyn SirTrace) -> Result<Self, InvalidTraceError> {
+    /// Create a TirTrace from a SirTrace, trimming remnants of the code which starts/stops the
+    /// tracer. Returns a TIR trace and the bounds the SIR trace was trimmed to, or Err if a DefId
+    /// is encountered for which no SIR is available.
+    pub fn new<'s>(trace: &'s dyn SirTrace) -> Result<Self, InvalidTraceError> {
         let mut ops = Vec::new();
-        let num_locs = trace.len();
+        let mut itr = trace.into_iter().peekable();
+        loop {
+            let loc = itr.next();
```
Comment (ltratt) [inline yktrace/src/tir.rs]: Isn't this just a `for` loop in disguise?

File: ykpack/src/types.rs (hunk 2 above)
Comment (vext01) [reply ykpack/src/types.rs]: In theory it could happen, but I doubt it. I've used a flags bitfield to allow for additional flags in the future.

File: yktrace/src/debug.rs (hunk 5 above)
Comment (vext01) [reply yktrace/src/debug.rs]: Because there may be more than one flag per trace location.

File: yktrace/src/lib.rs (hunk 6 above)
Comment (vext01) [reply yktrace/src/lib.rs]: Ah yes.

File: yktrace/src/debug.rs (hunk 5 above)
Comment (ltratt) [reply yktrace/src/debug.rs]: OK... but then still just do it as a series of `print`s after the one that's already there maybe?

File: yktrace/src/tir.rs (hunk 9 above)
Comment (vext01) [reply yktrace/src/tir.rs]: Yes, but the `peek()` later causes borrow check issues if you use a for loop.

File: yktrace/src/tir.rs (hunk 7 above)
Comment (ltratt) [reply yktrace/src/tir.rs]: Aha, I see what you mean. OK, maybe tweak a bit to make this clearer?

File: yktrace/src/tir.rs (hunk 9 above)
Comment (ltratt) [reply yktrace/src/tir.rs]: Ah, OK. So maybe:

```rust
loop {
    if let Some(loc) = itr.next() {
        ...
    } else {
        break;
    }
}
```

although that may also break the borrow checker... I'm just trying to avoid the conceptual and run-time overhead of `X.is_none(); ... ; X.unwrap()`.

File: yktrace/src/debug.rs (hunk 5 above)
Comment (vext01) [reply yktrace/src/debug.rs]: Yeah, I suppose we could do. I didn't put a great deal of effort into optimising this, as it's debug code that won't be called in production.

File: yktrace/src/tir.rs (hunk 7 above)
Comment (vext01) [reply yktrace/src/tir.rs]: Sure.

File: yktrace/src/debug.rs (hunk 5 above)
Comment (ltratt) [reply yktrace/src/debug.rs]: Understood. At the moment the intermediate list is just a bit of a distraction, so I think simplifying is a good idea.

File: yktrace/src/tir.rs (hunk 9 above)
Comment (vext01) [reply yktrace/src/tir.rs]: Your solution does work:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=77ecf18f4300a16e725651434c42825d

Whereas this is what I'd hoped to write:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ccf28d6a3d8635b051d65e554600b442

Comment (ltratt) [reply yktrace/src/tir.rs]: Ah, the borrow checker :)

Comment (bjorn3) [reply yktrace/src/tir.rs]: Have you tried `while let Some(loc) = iter.next() { ... }`?

Comment (vext01) [reply yktrace/src/tir.rs]: Indeed. Just got the same suggestion from someone on #rust IRC:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d6ee07b93af9c802f93c9eea076b3ebb

Comment (ltratt) [reply yktrace/src/tir.rs]: Good idea!

Comment (vext01) [reply yktrace/src/tir.rs]: Or even better, check this out:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4f03faceeef0f67cf30e7446f3b9e00b

Credit: M-ou-se on IRC :)

Comment (ltratt) [reply yktrace/src/tir.rs]: Does that work OK for the last element in the list?

Comment (vext01) [reply yktrace/src/tir.rs]: Ah, you are right. It's only gone around the loop twice. Shame.
```
[src/main.rs:5] i[0] = 1
[src/main.rs:6] i[1] = 2
[src/main.rs:5] i[0] = 2
[src/main.rs:6] i[1] = 3
```

**Comment by vext01** [comment]
I think that addresses the comments.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat, but we need to look at the other PR first.

**Comment by vext01** [comment]
OK, so the passing CI from ykrustc will have made new tarballs using the new ykpack and we can r+ this now.
